infrastructure/code/code.go

```diff
@@ -166,43 +175,80 @@ func (sc *Scanner) Scan(ctx context.Context, path string, folderPath string) (is
 		sc.changedFilesMutex.Unlock()
 		return []snyk.Issue{}, nil
 	}
-
-	changedFiles := make(map[string]bool)
-	for changedPath := range sc.changedPaths[folderPath] {
-		if !uri.IsDirectory(changedPath) {
-			changedFiles[changedPath] = true
-		}
-		delete(sc.changedPaths[folderPath], changedPath)
-	}
+	filesToBeScanned := sc.getFilesToBeScanned(folderPath)
 	sc.changedFilesMutex.Unlock()
 
 	startTime := time.Now()
 	span := sc.BundleUploader.instrumentor.StartSpan(ctx, "code.ScanWorkspace")
 	defer sc.BundleUploader.instrumentor.Finish(span)
 
 	// Start the scan
-	t := progress.NewTracker(false)
-	t.BeginWithMessage("Snyk Code: Collecting files in \""+folderPath+"\"", "Evaluating ignores and counting files...")
 	fileFilter, _ := sc.fileFilters.Load(folderPath)
 	if fileFilter == nil {
-		fileFilter = filefilter.NewFileFilter(folderPath, config.CurrentConfig().Logger())
+		fileFilter = filefilter.NewFileFilter(folderPath, &logger)
 		sc.fileFilters.Store(folderPath, fileFilter)
 	}
 	files := fileFilter.FindNonIgnoredFiles()
-	t.EndWithMessage("Collected files")
 	metrics := sc.newMetrics(startTime)
 
 	var results []snyk.Issue
 	if sc.useIgnoresFlow() {
-		results, err = sc.UploadAndAnalyzeWithIgnores(span.Context(), folderPath, files, changedFiles)
+		results, err = sc.UploadAndAnalyzeWithIgnores(span.Context(), folderPath, files, filesToBeScanned)
 	} else {
-		results, err = sc.UploadAndAnalyze(span.Context(), files, folderPath, metrics, changedFiles)
+		results, err = sc.UploadAndAnalyze(span.Context(), files, folderPath, metrics, filesToBeScanned)
 	}
-
+	sc.removeFromCache(filesToBeScanned)
+	sc.addToCache(results)
 	return results, err
 }
 
-func (sc *Scanner) waitForScanToFinish(scanStatus *ScanStatus, folderPath string) (waiting bool) {
+// getFilesToBeScanned returns a map of files that need to be scanned and removes them from the changedPaths set.
+// This function also analyzes interfile dependencies, taking into account the dataflow between files.
+func (sc *Scanner) getFilesToBeScanned(folderPath string) map[string]bool {
+	logger := config.CurrentConfig().Logger().With().Str("method", "code.getFilesToBeScanned").Logger()
+	changedFiles := make(map[string]bool)
+	for changedPath := range sc.changedPaths[folderPath] {
+		if uri.IsDirectory(changedPath) {
+			logger.Debug().Str("path", changedPath).Msg("skipping directory")
```

Comment:
I'll need to do a manual test but I noticed in the old code we would run `delete(sc.changedPaths[folderPath], changedPath)` regardless. I wonder if this makes a difference to anything

Reply:
I don't think it causes any real side effects actually. The only thing I could see happening is if there are multiple scans queued at once and an earlier scan goes through the `changedPaths`, the next scan would still need to see that the left over paths in `changedPaths` are directories and skip them.

Reply:
Yeah, it should not make a difference. Also, it's cleaner to have the guard first in the loop, I think.